rllib/agents/marwil/tests/test_marwil.py

```diff
@@ -70,6 +74,74 @@ def test_marwil_compilation_and_learning_from_offline_file(self):
 
             trainer.stop()
 
+    def test_marwil_loss_function(self):
+        """
+        To generate the historic data used in this test case, first run:
+        $ ./train.py --run=PPO --env=CartPole-v0 \
+          --stop='{"timesteps_total": 50000}' \
+          --config='{"output": "/tmp/out", "batch_mode": "complete_episodes"}'
+        """
+        rllib_dir = Path(__file__).parent.parent.parent.parent
+        print("rllib dir={}".format(rllib_dir))
+        data_file = os.path.join(rllib_dir, "tests/data/cartpole/small.json")
+        print("data_file={} exists={}".format(data_file,
+                                              os.path.isfile(data_file)))
+        config = marwil.DEFAULT_CONFIG.copy()
+        config["num_workers"] = 0  # Run locally.
+        # Learn from offline data.
+        config["input"] = [data_file]
+
+        for fw in framework_iterator(config, frameworks=["torch", "tf2"]):
+            reader = JsonReader(inputs=[data_file])
+            batch = reader.next()
+
+            trainer = marwil.MARWILTrainer(config=config, env="CartPole-v0")
+            policy = trainer.get_policy()
+            model = policy.model
+
+            # Calculate our own expected values (to then compare against the
+            # agent's loss output).
+            cummulative_rewards = compute_advantages(
+                batch, 0.0, config["gamma"], 1.0, False, False)["advantages"]
+            if fw == "torch":
+                cummulative_rewards = torch.tensor(cummulative_rewards)
+            tensor_batch = policy._lazy_tensor_dict(batch)
+            model_out, _ = model.from_batch(tensor_batch)
+            vf_estimates = model.value_function()
+            adv = cummulative_rewards - vf_estimates
+            if fw == "torch":
+                adv = adv.detach().cpu().numpy()
+            adv_squared = np.mean(np.square(adv))
+            c_2 = 100.0 + 1e-8 * (adv_squared - 100.0)
+            c = np.sqrt(c_2)
+            exp_advs = np.exp(config["beta"] * (adv / c))
+            logp = policy.dist_class(model_out,
+                                     model).logp(tensor_batch["actions"])
+            if fw == "torch":
+                logp = logp.detach().cpu().numpy()
+            # Calculate all expected loss components.
+            expected_vf_loss = 0.5 * adv_squared
```

Comment:
Isn't this already in MarwilPolicyLoss?

Reply:
Yes, but with this test (adding this simple numpy implementation) we make sure that all frameworks' MARWIL losses are always the same and we won't break anything in the future. We have the same tests for other algos offering more security against breaking losses.